File: speed/memory/memory.py
import sys
import time
import random
import threading
import gc
from concurrent.futures import ThreadPoolExecutor
import array

# simple arena allocator class
class Arena:
    def __init__(self, size):
        self.buffer = bytearray(size)
        self.used = 0
    
    def allocate(self, size):
        # align to 8 bytes
        size = (size + 7) & ~7
        
        if self.used + size > len(self.buffer):
            return None
        
        start = self.used
        self.used += size
        return memoryview(self.buffer[start:start + size])
    
    def reset(self):
        self.used = 0
    
    def capacity(self):
        return len(self.buffer)
    
    def usage(self):
        return self.used

# allocation patterns test - sequential, random, producer-consumer
def allocation_patterns_test(iterations):
    start = time.perf_counter()
    
    # sequential allocation pattern
    ptrs = []
    for i in range(iterations):
        size = 64 + (i % 256)
        ptrs.append(bytearray(size))
    
    ptrs.clear()
    gc.collect()
    
    # random allocation pattern
    random.seed(42)
    raw_ptrs = []
    
    for i in range(iterations):
        size = 32 + random.randint(0, 511)
        raw_ptrs.append(bytearray(size))
    
    # random deallocation (shuffle and clear)
    random.shuffle(raw_ptrs)
    raw_ptrs.clear()
    gc.collect()
    
    end = time.perf_counter()
    _ = iterations  # prevent optimization
    return (end - start) * 1000

# worker function for gc stress test
def gc_stress_worker(thread_id, iterations, counter_lock, counter):
    random.seed(42 + thread_id)
    
    for i in range(iterations):
        size = 16 + random.randint(0, 1023)
        data = bytearray(size)
        
        # simulate work
        for j in range(len(data)):
            data[j] = i & 0xFF
        
        # calculate sum
        total = 0
        for j in range(0, size, 8):
            total += data[j]
        _ = total  # prevent optimization
        
        with counter_lock:
            counter[0] += 1

# gc stress testing with multiple threads
def gc_stress_test(num_threads, iterations_per_thread):
    start = time.perf_counter()
    
    counter = [0]
    counter_lock = threading.Lock()
    
    with ThreadPoolExecutor(max_workers=num_threads) as executor:
        futures = []
        for i in range(num_threads):
            future = executor.submit(gc_stress_worker, i, iterations_per_thread, counter_lock, counter)
            futures.append(future)
        
        # wait for all threads to complete
        for future in futures:
            future.result()
    
    result = counter[0]
    _ = result  # prevent optimization
    
    end = time.perf_counter()
    return (end - start) * 1000

# cache locality and fragmentation test
def cache_locality_test(iterations):
    start = time.perf_counter()
    
    # allocate small and large objects interleaved
    small_ptrs = []
    large_ptrs = []
    
    random.seed(42)
    
    # interleaved allocation pattern
    for i in range(iterations):
        small_size = 16 + random.randint(0, 63)
        large_size = 1024 + random.randint(0, 4095)
        
        small_array = bytearray(small_size)
        large_array = bytearray(large_size)
        
        # access pattern to test spatial locality
        for j in range(len(small_array)):
            small_array[j] = i & 0xFF
        for j in range(min(1024, len(large_array))):
            large_array[j] = (i + 1) & 0xFF
        
        small_ptrs.append(small_array)
        large_ptrs.append(large_array)
    
    # random access pattern to stress cache
    for i in range(iterations // 2):
        idx1 = random.randint(0, iterations - 1)
        idx2 = random.randint(0, iterations - 1)
        
        if idx1 < len(small_ptrs):
            small_array = small_ptrs[idx1]
            total = 0
            for j in range(min(16, len(small_array))):
                total += small_array[j]
            _ = total
        
        if idx2 < len(large_ptrs):
            large_array = large_ptrs[idx2]
            total = 0
            for j in range(0, min(1024, len(large_array)), 64):
                total += large_array[j]
            _ = total
    
    end = time.perf_counter()
    return (end - start) * 1000

# memory pool performance test
def memory_pool_test(iterations):
    start = time.perf_counter()
    
    # test standard allocation
    std_ptrs = []
    for i in range(iterations):
        data = bytearray(128)
        for j in range(len(data)):
            data[j] = i & 0xFF
        std_ptrs.append(data)
    
    std_ptrs.clear()
    gc.collect()
    
    # test arena allocation
    arena = Arena(iterations * 128 + 1024)
    arena_ptrs = []
    
    for i in range(iterations):
        ptr = arena.allocate(128)
        if ptr is not None:
            for j in range(len(ptr)):
                ptr[j] = i & 0xFF
            arena_ptrs.append(ptr)
    
    # batch deallocation
    arena.reset()
    
    # test batch allocation
    for batch in range(10):
        for i in range(iterations // 10):
            ptr = arena.allocate(128)
            if ptr is not None:
                for j in range(len(ptr)):
                    ptr[j] = i & 0xFF
        arena.reset()
    
    end = time.perf_counter()
    return (end - start) * 1000

# memory intensive workloads test
def memory_intensive_test(large_size_mb):
    start = time.perf_counter()
    
    size = large_size_mb * 1024 * 1024
    
    # large object allocation
    large_array1 = bytearray(size)
    large_array2 = bytearray(size)
    
    # memory bandwidth test - sequential write
    for i in range(0, size, 4096):
        large_array1[i] = i & 0xFF
    
    # memory copy operations
    large_array2[:] = large_array1
    
    # memory bandwidth test - sequential read
    total = 0
    for i in range(0, size, 4096):
        total += large_array2[i]
    _ = total
    
    # memory access pattern test
    random.seed(42)
    for i in range(10000):
        offset = random.randint(0, size - 64)
        val = large_array1[offset]
        large_array2[offset] = (val + 1) & 0xFF
    
    end = time.perf_counter()
    return (end - start) * 1000

def main():
    scale_factor = 1
    
    if len(sys.argv) > 1:
        try:
            scale_factor = int(sys.argv[1])
            if scale_factor <= 0:
                scale_factor = 1
        except ValueError:
            print("Invalid scale factor. Using default 1.", file=sys.stderr)
            scale_factor = 1
    
    total_time = 0.0
    
    total_time += allocation_patterns_test(10000 * scale_factor)
    total_time += gc_stress_test(4, 2500 * scale_factor)
    total_time += cache_locality_test(5000 * scale_factor)
    total_time += memory_pool_test(8000 * scale_factor)
    total_time += memory_intensive_test(100 * scale_factor)
    
    print(f"{total_time:.3f}")

if __name__ == "__main__":
    main()

File: speed/memory/memory.java
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public class memory {
    
    // simple arena allocator class
    static class Arena {
        private byte[] buffer;
        private int used;
        
        public Arena(int size) {
            this.buffer = new byte[size];
            this.used = 0;
        }
        
        public byte[] allocate(int size) {
            // align to 8 bytes
            size = (size + 7) & ~7;
            
            if (used + size > buffer.length) {
                return null;
            }
            
            byte[] result = new byte[size];
            System.arraycopy(buffer, used, result, 0, size);
            used += size;
            return result;
        }
        
        public void reset() {
            used = 0;
        }
        
        public int capacity() { return buffer.length; }
        public int usage() { return used; }
    }
    
    // allocation patterns test - sequential, random, producer-consumer
    static double allocationPatternsTest(int iterations) {
        long start = System.nanoTime();
        
        // sequential allocation pattern
        List<byte[]> ptrs = new ArrayList<>(iterations);
        for (int i = 0; i < iterations; i++) {
            int size = 64 + (i % 256);
            ptrs.add(new byte[size]);
        }
        ptrs.clear();
        System.gc();
        
        // random allocation pattern
        Random rng = new Random(42);
        List<byte[]> rawPtrs = new ArrayList<>(iterations);
        
        for (int i = 0; i < iterations; i++) {
            int size = 32 + rng.nextInt(512);
            rawPtrs.add(new byte[size]);
        }
        
        // random deallocation (shuffle and clear)
        Collections.shuffle(rawPtrs, rng);
        rawPtrs.clear();
        System.gc();
        
        long end = System.nanoTime();
        return (end - start) / 1000000.0;
    }
    
    // worker class for gc stress test
    static class GcStressWorker implements Runnable {
        private final int threadId;
        private final int iterations;
        private final AtomicInteger counter;
        
        public GcStressWorker(int threadId, int iterations, AtomicInteger counter) {
            this.threadId = threadId;
            this.iterations = iterations;
            this.counter = counter;
        }
        
        @Override
        public void run() {
            Random rng = new Random(42 + threadId);
            
            for (int i = 0; i < iterations; i++) {
                int size = 16 + rng.nextInt(1024);
                byte[] data = new byte[size];
                
                // simulate work
                Arrays.fill(data, (byte)(i & 0xFF));
                
                byte sum = 0;
                for (int j = 0; j < size; j += 8) {
                    sum += data[j];
                }
                
                counter.incrementAndGet();
            }
        }
    }
    
    // gc stress testing with multiple threads
    static double gcStressTest(int numThreads, int iterationsPerThread) {
        long start = System.nanoTime();
        
        AtomicInteger counter = new AtomicInteger(0);
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        List<Future<?>> futures = new ArrayList<>();
        
        for (int i = 0; i < numThreads; i++) {
            futures.add(executor.submit(new GcStressWorker(i, iterationsPerThread, counter)));
        }
        
        try {
            for (Future<?> future : futures) {
                future.get();
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            executor.shutdown();
        }
        
        int result = counter.get();
        
        long end = System.nanoTime();
        return (end - start) / 1000000.0;
    }
    
    // cache locality and fragmentation test
    static double cacheLocalityTest(int iterations) {
        long start = System.nanoTime();
        
        // allocate small and large objects interleaved
        List<byte[]> smallPtrs = new ArrayList<>(iterations);
        List<byte[]> largePtrs = new ArrayList<>(iterations);
        
        Random rng = new Random(42);
        
        // interleaved allocation pattern
        for (int i = 0; i < iterations; i++) {
            int smallSize = 16 + rng.nextInt(64);
            int largeSize = 1024 + rng.nextInt(4096);
            
            byte[] smallArray = new byte[smallSize];
            byte[] largeArray = new byte[largeSize];
            
            // access pattern to test spatial locality
            Arrays.fill(smallArray, (byte)(i & 0xFF));
            Arrays.fill(largeArray, 0, Math.min(1024, largeArray.length), (byte)((i + 1) & 0xFF));
            
            smallPtrs.add(smallArray);
            largePtrs.add(largeArray);
        }
        
        // random access pattern to stress cache
        for (int i = 0; i < iterations / 2; i++) {
            int idx1 = rng.nextInt(iterations);
            int idx2 = rng.nextInt(iterations);
            
            if (idx1 < smallPtrs.size()) {
                byte[] smallArray = smallPtrs.get(idx1);
                byte sum = 0;
                for (int j = 0; j < Math.min(16, smallArray.length); j++) {
                    sum += smallArray[j];
                }
            }
            
            if (idx2 < largePtrs.size()) {
                byte[] largeArray = largePtrs.get(idx2);
                byte sum = 0;
                for (int j = 0; j < Math.min(1024, largeArray.length); j += 64) {
                    sum += largeArray[j];
                }
            }
        }
        
        long end = System.nanoTime();
        return (end - start) / 1000000.0;
    }
    
    // memory pool performance test
    static double memoryPoolTest(int iterations) {
        long start = System.nanoTime();
        
        // test standard allocation
        List<byte[]> stdPtrs = new ArrayList<>(iterations);
        for (int i = 0; i < iterations; i++) {
            byte[] data = new byte[128];
            Arrays.fill(data, (byte)(i & 0xFF));
            stdPtrs.add(data);
        }
        stdPtrs.clear();
        System.gc();
        
        // test arena allocation
        Arena arena = new Arena(iterations * 128 + 1024);
        List<byte[]> arenaPtrs = new ArrayList<>(iterations);
        
        for (int i = 0; i < iterations; i++) {
            byte[] ptr = arena.allocate(128);
            if (ptr != null) {
                Arrays.fill(ptr, (byte)(i & 0xFF));
                arenaPtrs.add(ptr);
            }
        }
        
        // batch deallocation
        arena.reset();
        
        // test batch allocation
        for (int batch = 0; batch < 10; batch++) {
            for (int i = 0; i < iterations / 10; i++) {
                byte[] ptr = arena.allocate(128);
                if (ptr != null) {
                    Arrays.fill(ptr, (byte)(i & 0xFF));
                }
            }
            arena.reset();
        }
        
        long end = System.nanoTime();
        return (end - start) / 1000000.0;
    }
    
    // memory intensive workloads test
    static double memoryIntensiveTest(int largeSizeMb) {
        long start = System.nanoTime();
        
        int size = largeSizeMb * 1024 * 1024;
        
        // large object allocation
        byte[] largeArray1 = new byte[size];
        byte[] largeArray2 = new byte[size];
        
        // memory bandwidth test - sequential write
        for (int i = 0; i < size; i += 4096) {
            largeArray1[i] = (byte)(i & 0xFF);
        }
        
        // memory copy operations
        System.arraycopy(largeArray1, 0, largeArray2, 0, size);
        
        // memory bandwidth test - sequential read
        long sum = 0;
        for (int i = 0; i < size; i += 4096) {
            sum += largeArray2[i];
        }
        
        // memory access pattern test
        Random rng = new Random(42);
        for (int i = 0; i < 10000; i++) {
            int offset = rng.nextInt(size - 64);
            byte val = largeArray1[offset];
            largeArray2[offset] = (byte)(val + 1);
        }
        
        long end = System.nanoTime();
        return (end - start) / 1000000.0;
    }
    
    public static void main(String[] args) {
        int scaleFactor = 1;
        
        if (args.length > 0) {
            try {
                scaleFactor = Integer.parseInt(args[0]);
                if (scaleFactor <= 0) scaleFactor = 1;
            } catch (NumberFormatException e) {
                System.err.println("Invalid scale factor. Using default 1.");
                scaleFactor = 1;
            }
        }
        
        double totalTime = 0.0;
        
        totalTime += allocationPatternsTest(10000 * scaleFactor);
        totalTime += gcStressTest(4, 2500 * scaleFactor);
        totalTime += cacheLocalityTest(5000 * scaleFactor);
        totalTime += memoryPoolTest(8000 * scaleFactor);
        totalTime += memoryIntensiveTest(100 * scaleFactor);
        
        System.out.printf("%.3f%n", totalTime);
    }
}